Reverse synstack before setting commentstring

JoosepAlviste opened the pull request:
I had a problem that the commentstring does not get set correctly in the following case in a JSX file (cursor noted by `|`):

```jsx
const MyComponent = () => {
    return {
        <button
            onClick={() => {
                |console.log('hello')
                 ^ Cursor is here
            }}
        >
            Hello!
        </button>
    }
}
```

The commentstring at that location should be `// %s` but it's not really possible to match the syntax group correctly. Here's how the synstack looks like at that location:

```
['jsFuncBlock', 'jsParen', 'jsxRegion', 'jsxTag', 'jsxEscapeJsAttributes', 'jsFuncBlock', 'jsGlobalObjects']
```

When looping through this list from left to right, the `jsxRegion` syntax group gets matched first and the commentstring will be `{/*%s*/}`. If the synstack is reversed, however, the first match will
be `jsFuncBlock` and the commentstring will be set to `// %s` as expected.

---

Do you think that this is a good solution for the problem? Maybe there's a better approach to get by example to work?

---

**Note:** I didn't test to see if the other file types work since I don't use those that much myself (`vim`, `html`, `xhtml`, `typescript.jsx`, `typescript.tsx`, `vue`).

JoosepAlviste committed:
Reverse synstack before setting commentstring

I had a problem that the commentstring does not get set correctly in the
following case in a JSX file (cursor noted by |):

```jsx
const MyComponent = () => {
    return {
        <button
            onClick={() => {
                |console.log('hello')
                 ^ Cursor is here
            }}
        >
            Hello!
        </button>
    }
}
```

The commentstring at that location should be `// %s` but it's not really
possible to match the syntax group correctly. Here's how the synstack
looks like at that location:

['jsFuncBlock', 'jsParen', 'jsxRegion', 'jsxTag',
'jsxEscapeJsAttributes', 'jsFuncBlock', 'jsGlobalObjects']

When looping through this list from left to right, the `jsxRegion`
syntax group gets matched first and the commentstring will be
`{/*%s*/}`. If the synstack is reversed, however, the first match will
be `jsFuncBlock` and the commentstring will be set to `// %s` as
expected.

suy commented:
Hello. Thank you for the patch. I did a quick try with the call to reverse added, and it seems to work for me on VimL+Lua and HTML+JS. Since JSX is not something I'm used to, can you give me a sample file to test a bit?

I've tried copying your snippet to a file named `.jsx`, and opening it the filetype gets set to `javascriptreact`, but I don't know how each area should be commented. I can imagine for `console.log` but that's about it. :)

JoosepAlviste commented:
Hey! Thanks for taking a look!

Here's a bigger example file:

```jsx
import React, { useState } from 'react';

function CounterExample() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button
        type="button"
        onClick={() => {
          setCount(count + 1)
        }}
      >
        Click me
      </button>
    </div>
  );
}

export default CounterExample;
```

The comments in JSX are quite weird. Normal JS comments work outside of JSX:

```jsx
// import React, { useState } from 'react';

// function CounterExample() {
  // const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button
        type="button"
        onClick={() => {
          setCount(count + 1)
        }}
      >
        Click me
      </button>
    </div>
  );
}

export default CounterExample;
```

But inside JSX (where the syntax group is `jsxRegion` for me), the comments need to be surrounded by curly braces:

```jsx
import React, { useState } from 'react';

function CounterExample() {
  const [count, setCount] = useState(0);

  return (
    <div>
      {/* <p>You clicked {count} times</p> */}
      <button
        type="button"
        onClick={() => {
          setCount(count + 1)
        }}
      >
        {/* Click me */}
      </button>
    </div>
  );
}

export default CounterExample;
```

If you want to comment out the whole `button` element, for example, then you would use the curly brace style of comments:

```jsx
import React, { useState } from 'react';

function CounterExample() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      {/* <button */}
      {/*   type="button" */}
      {/*   onClick={() => { */}
      {/*     setCount(count + 1) */}
      {/*   }} */}
      {/* > */}
      {/*   Click me */}
      {/* </button> */}
    </div>
  );
}

export default CounterExample;
```

However, if you want to comment out something inside an element's attributes, you would use the normal JS style:

```jsx
import React, { useState } from 'react';

function CounterExample() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button
        // type="button"  <-- This is something I can't even reliably replicate with `reverse`-ing the synstack (it's not a common case anyways)
        onClick={() => {
          // setCount(count + 1)  <-- This is now possible with reversing
        }}
      >
        Click me
      </button>
    </div>
  );
}

export default CounterExample;
```

For commenting the line with `setCount`, I added this configuration (also in this MR): `'jsFuncBlock' : '// %s', `. A few more examples I can think of with attributes that have values on multiple lines (functions are probably the most important though, imo, so I added the configuration to this MR):

```jsx
(
  <MyComponent
    objectProp={{
      key: 'value',
      // commented: 'example',
    }}
    arrayProp={[
        'value1',
        'value2',
        // 'commentedValue',
    ]}
    somethingElseCanAlsoBeOnANewLine={
      // 'a value here'
    }
    // andTheWholeAttributeCouldBeCommented
  />
)
```

And if you have a 1-line element, then you can use the `/* */` syntax as well (I don't think that this example is really relevant for this plugin though):

```jsx
(
  <button /* type="button" */ onClick={() => console.log('hello')>
    Hello
  </button>
)
```

Let me know if there's anything else I can help with!

---

Just a note: I use [`vim-jsx-improve`](https://github.com/neoclide/vim-jsx-improve) for JSX but it should be very similar to [`vim-jsx-pretty`](https://github.com/MaxMEllon/vim-jsx-pretty).

JoosepAlviste committed:
Dynamically set comments setting

The problem is that, in Vue files, the `comments` setting is set to
support HTML comments. This means that if you have enabled `r` in
`formatoptions` and you press <CR> while in a comment in the JS section,
then the comment character won't get auto-inserted.

Since the `commentstring` variable is already being set dynamically with
the same logic, then it makes sense to support the `comments` option
here as well.
